File: 2023/perl/day02a/day02a.pl
#!/usr/bin/env perl

use strict;
use warnings;

use constant TOTAL_RED   => 12;
use constant TOTAL_GREEN => 13;
use constant TOTAL_BLUE  => 14;

sub usage {
    print "usage: $0 <file>\n";
    exit 1;
}

sub process {
    my ($contents) = @_;
    my $result = 0;

    foreach my $line (split /\n/, $contents) {
        my ($game_part, $sets_part) = (split ': ', $line);
        my $game_id = int((split ' ', $game_part)[1]);
        my $valid = 1;

        foreach my $game_set (split '; ', $sets_part) {
            foreach my $draws (split ', ', $game_set) {
                my ($amount_str, $color) = split ' ', $draws;
		my $amount = int($amount_str);

                if ($color eq 'red' && $amount > TOTAL_RED) {
                    $valid = 0;
                }
                if ($color eq 'green' && $amount > TOTAL_GREEN) {
                    $valid = 0;
                }
                if ($color eq 'blue' && $amount > TOTAL_BLUE) {
                    $valid = 0;
                }
            }
        }

        if ($valid) {
            $result += $game_id;
        }
    }

    return $result;
}

sub main {
    if (@ARGV < 1) {
        usage();
    }

    my $filename = $ARGV[0];

    open my $infile, '<', $filename or die "read of input file '$filename' failed.\n";
    my $contents = do { local $/; <$infile> };
    close $infile;

    my $result = process($contents);
    print "result = $result\n";
}

if ($0 eq __FILE__) {
    main();
}

File: 2023/erlang/day02a/day02a.erl
-module (day02a).
-export ([main/1]).

total_red() -> 12.
total_green() -> 13.
total_blue() -> 14.

usage() ->
    Module = "day02a.erl",
    io:format("usage: escript ~p <file>~n", [Module]),
    halt(1).

process_lines([], Result) -> Result;
process_lines([Line | Rest], Result) ->
    {Valid, GameNum} = process_line(Line),
    case Valid of
        true -> process_lines(Rest, Result + GameNum);
        false -> process_lines(Rest, Result)
    end.

process_line(Line) ->
    GamePart = hd(string:split(Line, ": ")),
    GameNumPart = lists:last(string:split(GamePart, " ")),
    {GameNum, _} = string:to_integer(GameNumPart),
    DrawsPart = lists:last(string:split(Line, ": ")),
    Valid = lists:foldl(
        fun(Draws, CurrentValid) ->
            lists:foldl(
                fun(ColorAmount, DrawValid) ->
                    AmountStr = hd(string:split(ColorAmount, " ")),
                    Color = lists:last(string:split(ColorAmount, " ")),
                    {Amount, _} = string:to_integer(AmountStr),
                    case Color of
                        "red" -> DrawValid andalso (Amount =< total_red());
                        "green" -> DrawValid andalso (Amount =< total_green());
                        "blue" -> DrawValid andalso (Amount =< total_blue());
                        _ -> throw({error, "unexpected color"})
                    end
                end,
                CurrentValid, string:split(Draws, ", ", all)
            )
        end,
        true, string:split(DrawsPart, "; ", all)
    ),
    {Valid, GameNum}.

process_input(Contents) ->
    Trimmed = string:trim(Contents),
    Lines = string:split(Trimmed, "\n", all),
    process_lines(Lines, 0).

main([Filename]) ->
    {ok, File} = file:read_file(Filename),
    Contents = unicode:characters_to_list(File),
    Result = process_input(Contents),
    io:format("result = ~w~n", [Result]);
main(_) ->
    usage().
